enhancements/installer/kubernete-slb-owned-api-loadbalancers.md

```diff
@@ -0,0 +1,508 @@
+
+
+## Background
+
+Currently the bootstrapping happens in these three stages,
+
+**Stage 1:** Bootstrapping control plane on bootstrap host
+
+PUBLIC ZONE:api=public-lb-ip
+PRIVATE ZONE:api=private-lb-ip;api-int=private-lb-ip
+
+public-lb:bootstrap(green),master-0(red);master-1(red);master-2(red)
+private-lb:bootstrap(green);master-0(red);master-1(red);master-2(red)
+
+The `bootstrap` machine hosts the bootstrap-control-plane, and the kubelet on `master-{0,1,2}` use the private-lb to communicate with k8s API.
+
+**Stage 2:** Bootstrapping complete
+
+PUBLIC ZONE:api=public-lb-ip
+PRIVATE ZONE:api=private-lb-ip;api-int=private-lb-ip
+
+public-lb:bootstrap(red),master-0(green);master-1(green);master-2(green)
+private-lb:bootstrap(red);master-0(green);master-1(green);master-2(green)
+
+Bootstrapping of the k8s API is complete and therefore the bootstrap-control-plane has shut itself down on the `bootstrap` machine. Here the actual 
+control-plane is serving all the request for k8s API
+
+**Stage 3:** Bootstrap host removed
+
+PUBLIC ZONE:api=public-lb-ip
+PRIVATE ZONE:api=private-lb-ip;api-int=private-lb-ip
+
+public-lb:master-0(green);master-1(green);master-2(green)
+private-lb:master-0(green);master-1(green);master-2(green)
+
+The installer terminates the `bootstrap` machine and also removes the `bootstrap` machine from the backend of all the loadbalancers. And only 
+the `master-{0,1,2}` machines remain.
+
+During all these stages, all the clients end up using the loadbalancers as a single abstraction and do not have to understand the bootstrapping 
+stages or even notice the bootstrap-control-plane.
+
+## Proposal
+
+During bootstrapping the clients use the known API, and API-INT DNS names to discover the endpoints. Ensuring that the clients continue to depend on 
+that abstraction is very important as it allows the clients to be simple.
+
+Secondly, the load balancers cannot be created by Kubernetes cloud controller manager unless there is a Kubernetes API already running, but for 
+successful bootstrapping we need the loadbalancers so that the kubelets can communicate with bootstrap-control-plane to run the necessary pods to create 
+the k8s API.
+
+So, the break the cyclic dependency, I suggest we modify the stages as follows,
+
+**Stage 1:** Bootstrapping control plane on bootstrap host
+
+PUBLIC ZONE:api=static-public-lb-ip
+PRIVATE ZONE:api=bootstrap-private-ip,static-private-lb-ip;api-int=bootstrap-private-ip,static-private-lb-ip
+
+The `bootstrap` machine hosts the bootstrap-control-plane, and the installer has pre-seeded the DNS entries with reserved ips for loadbalancers to 
+be created by the cluster. The kubelet on `master-{0,1,2}` use the DNS to resolve the endpoints to the `bootstrap` machine, 
+and communicate with the k8s API.
+
+An entity generates Kubernetes service type loadbalancer objects using the static ips created by the installer and pushes them to the cluster using 
```

Comment:
this "an entity" is the crucial bit. It would be natural this to be the same entity that does slb management in a running cluster. We don't want to duplicate this existing code.

Nothing speaks against running necessary components as part of the bootstrap control plane, i.e. in render mode and then as static pod on bootstrap. But I understand that we don't even need that:
- the LB-INT will be working with the pre-seeded DNS entries
- so kubelet will launch on `master-{0,1,2,}` and the slb backend (cloud-cm?) will be able to come up on the masters
- ... and replace the pre-seeded DNS entries with a real cloud LB.

@frobware does that sound right?

Reply:
Seeing kube-controller-manager one line below. Is it really kcm today? Or ccm or even something else?